src/BullOak.Repositories.EventStore/EventStoreSession.cs

```diff
@@ -0,0 +1,161 @@
+                            }).ToArray(),
+                            currentVersion);
+
+        }
+
+        private void CheckDisposedState()
+        {
+            if (isInDisposedState)
+            {
+                //this is purely design decision, nothing prevents implementing the session that support any amount and any order of oeprations
+                throw new InvalidOperationException("EventStoreSession should not be used after SaveChanges call");
+            }
+        }
+
+        private async Task<List<ResolvedEvent>> ReadAllEventsFromStream()
+        {
+            checked
+            {
+                var result = new List<ResolvedEvent>();
+                StreamEventsSlice currentSlice;
+                long nextSliceStart = StreamPosition.Start;
+                do
+                {
+                    currentSlice = await eventStoreConnection.ReadStreamEventsForwardAsync(streamName, nextSliceStart, SliceSize, false);
+                    nextSliceStart = currentSlice.NextEventNumber;
+                    result.AddRange(currentSlice.Events);
+                } while (!currentSlice.IsEndOfStream);
+                currentVersion = (int)currentSlice.LastEventNumber;
+                return result;
+            }
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing) { ConsiderSessionDisposed(); }
+            base.Dispose(disposing);
+        }
+
+        private void ConsiderSessionDisposed()
+        {
+            isInDisposedState = true;
+        }
+
+        /// <summary>
+        /// Saves changes to the respective stream
+        /// NOTES: Current implementation doesn't support cancellation token
+        /// </summary>
+        /// <param name="eventsToAdd"></param>
+        /// <param name="shouldSaveSnapshot"></param>
+        /// <param name="snapshot"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        protected override async Task<int> SaveChanges(object[] eventsToAdd,
+            bool shouldSaveSnapshot,
+            TState snapshot,
+            CancellationToken? cancellationToken)
+        {
+            checked
+            {
+                CheckDisposedState();
+                var writeResult = await eventStoreConnection.ConditionalAppendToStreamAsync(
+                    streamName,
+                    currentVersion,
+                    eventsToAdd.Select(eventObject =>
+                    {
+                        return CreateEventData(eventObject);
+                    }))
+                    .ConfigureAwait(false);
+
+                switch (writeResult.Status)
+                {
+                    case ConditionalWriteStatus.Succeeded:
+                        break;
+                    case ConditionalWriteStatus.VersionMismatch:
+                        throw new InvalidOperationException($"Stream expected version mismatched actual version. StreamId: {streamName}");
```

Comment:
This is the definition of a concurrency exception and should throw a concurrency exception rather than an `InvalidOperationException`